skynet/dgpu/daemon.py

```diff
@@ -32,274 +29,182 @@ def convert_reward_to_int(reward_str):
-    # -[ ] avoids so much indentation due to methods
-    async def serve_forever(self):
+        logging.warning('model not blacklisted!, skip...')
+        return
+
+    results = [res['request_id'] for res in conn._tables['results']]
+
+    # if worker already produced a result for this request
+    if rid in results:
+        logging.info(f'worker already submitted a result for request #{rid}, skip...')
+        return
+
+    statuses = conn._tables['requests'][rid]
+
+    # skip if workers in non_compete already on it
+    competitors = set((status['worker'] for status in statuses))
+    if bool(config.non_compete & competitors):
+        logging.info('worker in configured non_compete list already working on request, skip...')
+        return
+
+    # resolve the ipfs hashes into the actual data behind them
+    inputs = []
+    raw_inputs = req['binary_data'].split(',')
+    if raw_inputs:
+        logging.info(f'fetching IPFS inputs: {raw_inputs}')
+
+    retry = 3
+    for _input in req['binary_data'].split(','):
+        if _input:
+            for r in range(retry):
+                try:
+                    # user `GPUConnector` to IO with
+                    # storage layer to seed the compute
+                    # task.
+                    img = await conn.get_input_data(_input)
+                    inputs.append(img)
+                    logging.info(f'retrieved {_input}!')
+                    break
+
+                except BaseException:
+                    logging.exception(
+                        f'IPFS fetch input error !?! retries left {retry - r - 1}\n'
+                    )
+
+    # compute unique request hash used on submit
+    hash_str = (
+        str(req['nonce'])
+        +
+        req['body']
+        +
+        req['binary_data']
+    )
+    logging.debug(f'hashing: {hash_str}')
+    request_hash = sha256(hash_str.encode('utf-8')).hexdigest()
+    logging.info(f'calculated request hash: {request_hash}')
+
+    params = body['params']
+    total_step = params['step'] if 'step' in params else 1
+    model = body['params']['model']
+    mode = body['method']
+
+    # TODO: validate request
+
+    resp = await conn.begin_work(rid)
+    if not resp or 'code' in resp:
+        logging.info('begin_work error, probably being worked on already... skip.')
+        return
+
+    with maybe_load_model(model, mode) as model:
         try:
-            while True:
-                if self.auto_withdraw:
-                    await self.conn.maybe_withdraw_all()
+            maybe_update_tui(lambda tui: tui.set_progress(0, done=total_step))
+
+            output_type = 'png'
+            if 'output_type' in body['params']:
+                output_type = body['params']['output_type']
+
+            output = None
+            output_hash = None
+            match config.backend:
+                case 'sync-on-thread':
+                    output_hash, output = await trio.to_thread.run_sync(
```

Comment:
a concurrency architecture comment here detailing how this is where we block indefinitely until a model generates a result AND that the `trio.from_thread` call above is used to unblock other tasks from running in this parent thread since, as mentioned this should block until the work request is completed right?